---
title: "Window based HVF calling on residual overdispersion"
author: "C.A.Kapourani & R. Argelaguet"
output: 
  html_notebook:
    df_print: paged
    highlight: haddock
    number_sections: yes
    theme: cerulean
    toc: yes
---

```{r}
# Define ggplot2 theme for scatter plots
.scatter_theme <- function(legend_pos = "top") {
  p <- theme(
    plot.title = element_text(hjust = 0.5, size = rel(1.1),
                              margin = margin(0,0,2,0), color = "black"),
    legend.position = legend_pos,
    legend.title = element_blank(),
    axis.line = element_line(),
    #panel.border = element_blank(),
    #panel.grid.major = element_line(colour = "gainsboro"),
    panel.background = element_blank(),
    axis.text = element_text(color = "black", size = rel(0.8)),
    axis.title = element_text(color = "black", size = rel(1.2))
  )
  return(p)
}

scatter_plot <- function(df, x = "mu", y = "gamma", 
                         x_lab = expression(paste("Mean methylation ", mu)), 
                         y_lab = expression(paste("Overdispersion ", gamma)),
                         main = "Coloured by 'scMET' HVFs") {
  task <- "HVF"
  task <- toupper(task)
  size <- c(1.6, 0.65)
  names(size) <- c(task, "Other")
  fill <- c("red", "gray80")
  names(fill) <- c(task, "Other")
  alpha <- c(0.65, 0.45)
  names(alpha) <- c(task, "Other")
  
  gg <- ggplot(df, aes_string(x = x, y = y)) +
    geom_point(aes(fill = ifelse(is_variable, task, "Other"),
                   size = ifelse(is_variable, task, "Other"),
                   alpha = ifelse(is_variable, task, "Other")),
                   colour = "black", shape = 21, stroke = 0.03) +
    scale_fill_manual(values = fill) +
    scale_size_manual(values = size) +
    scale_alpha_manual(values = alpha) +
    theme_classic()  + 
    xlab(x_lab) + ylab(y_lab) +
    ggtitle(main) +
    .scatter_theme(legend_pos = "right")
  return(gg)
}
```


# Load libraries and settings
```{r}
suppressPackageStartupMessages(library(scMET))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(ggpubr))

## I/O
source("../../load_settings.R")
data_dir <- "~/datasets/scMET_ms/ecker2017/all_cells/data/"
marker_genes_file <- "~/datasets/scMET_ms/ecker2017/metadata/ecker2017_marker_genes.csv"
feat2genes_file <- paste0(io$basedir, "/features/genes2features100k.txt.gz")
out_dir <- "~/datasets/scMET_ms/ecker2017/all_cells/hvf_window/"
hits_dir <- "~/datasets/scMET_ms/ecker2017/all_cells/hvf_window/hits/"
if (!dir.exists(out_dir)) { dir.create(out_dir, recursive = TRUE) }
if (!dir.exists(hits_dir)) { dir.create(hits_dir, recursive = FALSE) }

## Options
marker_feat <- c("window20000_step20000_2848", 
                 "window20000_step20000_3440")
mode <- "epsilon"
chr <- paste0("chr", c(seq(1, 19), "X"))
# Define genomic contexts
annot <- list(
  c("window10000_step10000" = "10kb Windows"),
  c("window20000_step20000" = "20kb Windows")
)
```

# Load metadata information
```{r}
# Marker genes
marker_genes <- fread(marker_genes_file)
# Mapping features to nearest genes
feat2genes <- fread(io$features2genes)
```

# HVF analysis
Load HVF results from each chromosome and combine them together. From the mean-overdispersion plots we observe the same pattern across chromosomes, hence we just combine the HVF results and sort by variability.
```{r}
hvf_summary <- list()
for (i in 1:length(annot)) {
  chr_summary <- list()
  for (c in chr) {
    chr_summary[[c]] <- fread(sprintf("%s/hvf_%s_%s_%s.txt.gz", 
                                      hits_dir, names(annot[[i]]), c, mode))
  }
  hvf_summary[[annot[[i]]]] <- rbindlist(chr_summary) %>% setorder(-tail_prob, -epsilon)
  #fwrite(hvf_summary[[i]], 
  #       file = sprintf("%s/hvf_%s_%s.txt.gz", hits_dir, names(annot[[i]]), mode))
}
```


# Plot mean - overdispersion relationship coloured by HVF
Here we plot only for 20Kb windows
```{r}
set.seed(123)
an <- 2
##############
## Add labels
##############
hv <- hvf_summary[[annot[[an]]]]

# Features that are non HVF/LVF
tmp <- hv[hv$is_variable == FALSE, ]
tmp <- tmp[sample(NROW(tmp), 4000), ]
hv <- rbind(hv[hv$is_variable == TRUE, ], tmp)

# Store final labels
lab <- hv[is_variable == TRUE & id %in% marker_feat] %>%
  setorder(-epsilon) 
lab <- cbind(lab, data.table(label = c("window1", "window2")))
# Total number of hits
sig_hits <-  hv[is_variable == TRUE, .N]
nonsig_hits <-  hv[is_variable == FALSE, .N]
ylim <- max(hv$gamma, na.rm = TRUE) + 0.1
gg <- scatter_plot(df = hv, x = "mu", y = "gamma", main = NULL) +
  theme(legend.position = c(0.9, 0.92)) +
  geom_text_repel(force = 1, data = lab, aes_string(label = "label"),
                  size = 3, col = "black", segment.color = "black", 
                  segment.size = 0.3, segment.alpha = 0.35, 
                  box.padding = unit(0.5,"lines"), show.legend = FALSE
)
print(gg)

png(sprintf("%s/label_ecker_hvf_%s_%s.png", out_dir, names(annot[[an]]), 
            mode), res = 150, width = 800, height = 600)
print(gg)
dev.off()

pdf(sprintf("%s/label_ecker_hvf_%s_%s.pdf", out_dir, names(annot[[an]]), 
            mode), width = 8, height = 4)
print(gg)
dev.off()
```


# Enrichment test of HVFs being marker genes
```{r, fig.width=2, fig.height=2}
set.seed(123)
gg_list <- list()
df_enrich_list <- list()
i <- 2 # Only show plot for 20kb windows
#for (i in 1:length(annot)) {
hv <- hvf_summary[[annot[[i]]]] %>%  
  merge(feat2genes[,c("id","anno","gene")], by = c("id","anno"))
scmet_enrich <- hv[is_variable == TRUE & 
                     gene %in% marker_genes$gene_name] %>% .[, .N]
total_hits <- hv[is_variable == TRUE] %>% .[, .N]
null_distr <- c()
dt_copy <- copy(hv)
for (iter in 1:1000) {
  idx <- sample(NROW(hv), total_hits)
  null_distr[iter] <- dt_copy[, ran_hvf := FALSE] %>% 
    .[idx, ran_hvf := TRUE] %>%
    .[ran_hvf == TRUE & gene %in% marker_genes$gene_name] %>% .[, .N]
}

df_enrich <- data.frame(x = null_distr)
df_enrich_list[[i]] <- df_enrich
gg <- ggplot(df_enrich, aes(x = x)) +
  geom_histogram(aes(y = ..density..), position = "identity", 
                 fill = "grey75", color = "grey65", bins = 20) +
  geom_vline(xintercept = scmet_enrich, color = "#E69F00", 
             linetype = "dashed", size = 2) +
  xlab("Enrichment null distribution") + 
  ggtitle(annot[[i]]) + ylab("Density") +
  theme_classic() +
  theme(
    legend.position = c(0.85, 0.167),
    plot.title = element_text(hjust = 0.5),
    axis.text = element_text(color = "black", size = rel(0.8)),
    axis.title = element_text(color = "black", size = rel(1.1))
  )
#gg_list[[i]] <- gg
#}

print(gg)
pdf(sprintf("%s/ecker_enrichment_hvf_window20kb_%s.pdf", out_dir, mode),
    width = 4.5, height = 3.5, useDingbats = FALSE)
print(gg)
dev.off()
```
